controllers/component/component_adapter.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package component
+
+import (
+	"context"
+	"fmt"
+
+	//"reflect"
+
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/gitops"
+	h "github.com/redhat-appstudio/integration-service/helpers"
+	"github.com/redhat-appstudio/integration-service/loader"
+	"github.com/redhat-appstudio/integration-service/metrics"
+	"github.com/redhat-appstudio/integration-service/status"
+	"github.com/redhat-appstudio/operator-toolkit/controller"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a integration PipelineRun.
+type Adapter struct {
+	component   *applicationapiv1alpha1.Component
+	application *applicationapiv1alpha1.Application
+	loader      loader.ObjectLoader
+	logger      h.IntegrationLogger
+	client      client.Client
+	context     context.Context
+	status      status.Status
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(component *applicationapiv1alpha1.Component, application *applicationapiv1alpha1.Application, logger h.IntegrationLogger, loader loader.ObjectLoader, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		component:   component,
+		application: application,
+		logger:      logger,
+		loader:      loader,
+		client:      client,
+		context:     context,
+		status:      status.NewAdapter(logger.Logger, client),
+	}
+}
+
+// EnsureComponentIsCleanedUp is an operation that will ensure components
+// marked for deletion have a snapshot created without said component
+func (a *Adapter) EnsureComponentIsCleanedUp() (controller.OperationResult, error) {
+	if !HasComponentBeenDeleted(a.component) {
+		return controller.ContinueProcessing()
+	}
+
+	applicationComponents, err := a.loader.GetAllApplicationComponents(a.client, a.context, a.application)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	// TODO broken here
+	//var updatedComponents = []applicationapiv1alpha1.Component{}
+	comp := *(a.component)
+
+	/* 	for i, individualComponent := range *applicationComponents {
+	   		if comp.ObjectMeta.UID == individualComponent.ObjectMeta.UID {
+	   			//if reflect.DeepEqual(comp, individualComponent) {
```

Comment:
I don't think that using `DeepEqual` here is needed - I think that checking the `name` should be enough for this kind of comparison.

Reply:
Disregard, the approach that's below works even better.

Reply:
@dirgim yeah the deep equal here was causing some grief. I had seen it[ used here](https://github.com/redhat-appstudio/integration-service/blob/d2f5e27de0f49866fe8d61036ad27c22a3496a8d/gitops/snapshot.go#L389) to compare components. I used UID for this patch because I was wondering if there might be a situation where a deleted component might share the name of another vaild component within the application. In the case that that a component has been added to an application to replace the component with a different version/image.